### PR TITLE
Add instructions and processor for stake deactivate and withdraw

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Missing instruction for deactivating stake

#### Summary of Changes
Added code and tests

#4357

### CODE REVIEW

**Comment by pgarg66** [comment]
@rob-solana

File: programs/stake_api/src/stake_instruction.rs [hunk 1]
```diff
@@ -31,6 +31,22 @@ pub enum StakeInstruction {
     ///    2 - RewardsPool Stake Account from which to redeem credits
     ///    3 - Rewards syscall Account that carries points values
     RedeemVoteCredits,
+
+    /// Deactivates the stake in the account
+    ///
+    /// Expects 1 Accounts:
+    ///    0 - Delegate StakeAccount
+    Deactivate,
```
Comment (rob-solana) [inline programs/stake_api/src/stake_instruction.rs]: one of these at a time

File: programs/stake_api/src/stake_instruction.rs [hunk 2]
```diff
@@ -31,6 +31,22 @@ pub enum StakeInstruction {
     ///    2 - RewardsPool Stake Account from which to redeem credits
     ///    3 - Rewards syscall Account that carries points values
     RedeemVoteCredits,
+
+    /// Deactivates the stake in the account
+    ///
+    /// Expects 1 Accounts:
+    ///    0 - Delegate StakeAccount
```
Comment (rob-solana) [inline programs/stake_api/src/stake_instruction.rs]: need 2: the delegate and syscall::current

File: programs/stake_api/src/stake_instruction.rs [hunk 3]
```diff
@@ -123,6 +139,24 @@ pub fn process_instruction(
                 &syscall::rewards::from_keyed_account(&rest[0])?,
             )
         }
+        StakeInstruction::Deactivate => {
+            if rest.len() != 0 {
+                Err(InstructionError::InvalidInstructionData)?;
+            }
+            me.deactivate_stake(&syscall::current::from_keyed_account(&me)?)
```
Comment (rob-solana) [inline programs/stake_api/src/stake_instruction.rs]: rest.len() should be exactly 1 and current comes from &rest[0]

File: programs/stake_api/src/stake_state.rs [hunk 4]
```diff
@@ -276,6 +282,33 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
...
+        if let StakeState::Stake(stake) = self.state()? {
+            // Assume full stake if the stake is under warmup/cooldown
+            let staked = if stake.stake(current.epoch) == 0 {
+                0
+            } else {
+                stake.stake
+            };
+            if lamports > self.account.lamports.saturating_sub(staked) {
+                return Err(InstructionError::InsufficientFunds);
+            }
```
Comment (rob-solana) [inline programs/stake_api/src/stake_state.rs]: I think we should also lower stake.stake to whatever remains in the account

File: programs/stake_api/src/stake_instruction.rs (hunk 1 above)
Comment (pgarg66) [reply programs/stake_api/src/stake_instruction.rs]: 👍

File: programs/stake_api/src/stake_instruction.rs (hunk 2 above)
Comment (pgarg66) [reply programs/stake_api/src/stake_instruction.rs]: 👍

File: programs/stake_api/src/stake_instruction.rs (hunk 3 above)
Comment (pgarg66) [reply programs/stake_api/src/stake_instruction.rs]: 👍

File: programs/stake_api/src/stake_state.rs (hunk 4 above)
Comment (pgarg66) [reply programs/stake_api/src/stake_state.rs]: 👍

File: programs/stake_api/src/stake_instruction.rs [hunk 5]
```diff
@@ -284,6 +277,41 @@ mod tests {
             ),
             Err(InstructionError::InvalidAccountData),
         );
+
+        // gets the deserialization checks in deactivate
```
Comment (rob-solana) [inline programs/stake_api/src/stake_instruction.rs]: make sure comments are correct

**Comment by rob-solana** [comment]
these can land on master now

**Pull request closed without merging**